simtfl/bft/__init__.py

```diff
@@ -0,0 +1,142 @@
+"""
+Abstractions for Byzantine Fault-Tolerant protocols.
+
+The model of a BFT protocol assumed here was written for the purpose
+of Crosslink, based on the example of adapted-Streamlet ([CS2020] as
+modified in [Crosslink]). It might not be sufficient for other BFT
+protocols — but that's okay; it's a prototype.
+
+[CS2020] https://eprint.iacr.org/2020/088.pdf
+[Crosslink] https://hackmd.io/JqENg--qSmyqRt_RqY7Whw?view
+"""
+
+
+def two_thirds_threshold(n):
+    """
+    Calculate the notarization threshold used in most permissioned BFT protocols:
+    `ceiling(n * 2/3)`.
+    """
+    return (n * 2 + 2) // 3
+
+
+class PermissionedBFTBase:
+    """
+    This class is used for the genesis block in a permissioned BFT protocol
+    (which is taken to be notarized, and therefore valid, by definition).
+
+    It is also used as a base class for other BFT block and proposal classes.
+    """
+    def __init__(self, n, t):
+        """
+        Constructs a genesis block for a permissioned BFT protocol with
+        `n` nodes, of which at least `t` must sign each proposal.
+        """
+        self.n = n
+        self.t = t
+        self.parent = None
+
+
+class PermissionedBFTBlock(PermissionedBFTBase):
+    """
+    A block for a BFT protocol. Each non-genesis block is based on a
+    notarized proposal.
+
+    BFT blocks are taken to be notarized, and therefore valid, by definition.
+    """
+
+    def __init__(self, proposal):
+        super().__init__(proposal.n, proposal.t)
+
+        assert proposal.is_notarized()
+        self.proposal = proposal
+
+
+class PermissionedBFTProposal(PermissionedBFTBase):
+    """A proposal for a BFT protocol."""
+
+    def __init__(self, parent):
+        """
+        Constructs a `PermissionedBFTProposal` with the given parent
+        `PermissionedBFTBlock`. The parameters are determined by the parent
+        block.
+        """
+        super().__init__(parent.n, parent.t)
+        self.parent = parent
+        self.signers = set()
+
+    def assert_valid(self):
+        """
+        Assert that this proposal is valid. This does not assert that it is
+        notarized. This should be overridden by subclasses.
+        """
+        pass
+
+    def is_valid(self):
+        """Is this proposal valid?"""
+        try:
+            self.assert_valid()
+            return True
+        except AssertionError:
+            return False
+
+    def assert_notarized(self):
+        """
+        Assert that this proposal is notarized. A `PermissionedBFTProposal`
+        is notarized iff it is valid and has at least the threshold number of
+        signatures.
+        """
+        assert self.is_valid()
```

Comment:
```suggestion
        self.assert_valid()
```
(to give better error reporting).